### PR TITLE
HSEARCH-4069 Measure time more precisely when computing execution time of Lucene queries

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/HSEARCH-4069

### CODE REVIEW

File: engine/src/main/java/org/hibernate/search/engine/search/timeout/spi/TimeoutManager.java [hunk 1]
```diff
@@ -138,12 +141,18 @@ public boolean hasHardTimeout() {
...
+	 * This method may be performance expensive.
+	 *
+	 * @return high precision duration of took time.
+	 */
 	public Duration tookTime() {
-		return Duration.ofMillis( elapsedTimeInMilliseconds() );
+		return Duration.ofMillis( timingSource.nanoTime() - nanoTimeStart );
 	}
 
 	protected long elapsedTimeInMilliseconds() {
```
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/search/engine/search/timeout/spi/TimeoutManager.java:154]: Maybe the reason for having two methods with a similar purpose would be more obvious if you renamed this to `elapsedTimeEstimateMillis()`?

File: engine/src/main/java/org/hibernate/search/engine/search/timeout/spi/TimeoutManager.java [hunk 2]
```diff
@@ -79,11 +80,13 @@ public TimeoutManager(TimingSource timingSource, Long timeoutValue, TimeUnit tim
 	 * we start counting from this method call (if needed)
 	 */
 	public void start() {
-		this.start = timingSource.monotonicTimeEstimate();
+		this.monotonicTimeEstimateStart = timingSource.monotonicTimeEstimate();
```
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/search/engine/search/timeout/spi/TimeoutManager.java:83]: Actually, if no timeout was defined, this line becomes unnecessary.

That in itself wouldn't be worth discussing, but... if you only execute this line when there actually is a timeout, then you can do the same for `timingSource.ensureInitialized();`, in the constructor above. And *that* method can be costly, as it triggers the creation of a thread pool on its first execution.

**Review by yrodiere — COMMENTED** [review]
Looks good! Just a few suggestions.

File: engine/src/main/java/org/hibernate/search/engine/search/timeout/spi/TimeoutManager.java (hunk 1 above)
Comment (fax4ever) [reply engine/src/main/java/org/hibernate/search/engine/search/timeout/spi/TimeoutManager.java:154]: Good idea!

File: engine/src/main/java/org/hibernate/search/engine/search/timeout/spi/TimeoutManager.java (hunk 2 above)
Comment (fax4ever) [reply engine/src/main/java/org/hibernate/search/engine/search/timeout/spi/TimeoutManager.java:83]: Right, we don't need it anymore to compute the took time!

**Comment by fax4ever** [comment]
Thanks for the reviews and the ideas. I did some changes.

**Comment by yrodiere** [comment]
Merged, thanks!

**Pull request merged**